mu-plugins/central-hub/tests/phpunit/integration/meta-data/getCustomFieldsValues.php

```diff
@@ -0,0 +1,151 @@
+	}
+
+	/**
+	 * Test get_custom_fields_values() should return an empty array when custom fields config is empty.
+	 */
+	public function test_should_return_empty_array_when_custom_fields_config_is_empty() {
+		$config = [
+			'custom_fields' => [],
+		];
+
+		$this->assertSame( [], get_custom_fields_values( $this->post, 'events', $config ) );
+	}
+
+	/**
+	 * Test get_custom_fields_values() should return post meta from database when meta key exists in custom fields
+	 * config.
+	 */
+	public function test_should_return_post_meta_from_database_when_meta_key_exists_in_custom_fields_config() {
+		$config = [
+			'custom_fields' => [
+				'event-date' => [
+					'is_single' => true,
+					'default'   => '',
+				],
+				'event-time' => [
+					'is_single' => true,
+					'default'   => '',
+				],
+				'venue-name' => [
+					'is_single' => true,
+					'default'   => '',
+				],
+			],
+		];
+
+		// Add post meta to the database so we have something to call.
+		add_post_meta( $this->post, 'event-date', '10-04-2019' );
+		add_post_meta( $this->post, 'event-time', '19:30:00' );
+		add_post_meta( $this->post, 'venue-name', 'Carnegie Hall' );
+
+		// Check database that post meta was added.
+		$this->assertSame( '10-04-2019', get_post_meta( $this->post, 'event-date', true ) );
+		$this->assertSame( '19:30:00', get_post_meta( $this->post, 'event-time', true ) );
+		$this->assertSame( 'Carnegie Hall', get_post_meta( $this->post, 'venue-name', true ) );
+
+		$expected_custom_fields = [
+			'event-date' => '10-04-2019',
+			'event-time' => '19:30:00',
+			'venue-name' => 'Carnegie Hall',
+		];
+
+		$this->assertSame( $expected_custom_fields, get_custom_fields_values( $this->post, 'events', $config ) );
+
+		// Clean up database.
+		delete_post_meta( $this->post, 'event-date', '10-04-2019' );
+		delete_post_meta( $this->post, 'event-time', '19:30:00' );
+		delete_post_meta( $this->post, 'venue-name', 'Carnegie Hall' );
+	}
+
+	/**
+	 * Test get_custom_fields_values() should return custom fields default values when post meta is not in database.
+	 */
+	public function test_should_return_custom_fields_default_values_when_post_meta_is_not_in_database() {
+		$config = [
+			'custom_fields' => [
+				'event-date' => [
+					'is_single' => true,
+					'default'   => '',
+				],
+				'event-time' => [
+					'is_single' => true,
+					'default'   => '',
+				],
+				'venue-name' => [
+					'is_single' => true,
+					'default'   => '',
+				],
+			],
+		];
+
+		// Check that database does not contain post meta.
+		$this->assertSame( '', get_post_meta( $this->post, 'event-date', true ) );
+		$this->assertSame( '', get_post_meta( $this->post, 'event-time', true ) );
+		$this->assertSame( '', get_post_meta( $this->post, 'venue-name', true ) );
+
+		$expected_custom_fields = [
+			'event-date' => '',
+			'event-time' => '',
+			'venue-name' => '',
+		];
+
+		$this->assertSame( $expected_custom_fields, get_custom_fields_values( $this->post, 'events', $config ) );
+	}
+}
+
```

Comment:
@rgadon107 What about the `'filter_meta_box_custom_fields '` filter? You'll want to test that:

1. it fires
2. the expected arguments are received.

Reply:
@hellofromtonya I can prove item 1 above. Not sure how to execute 2 above. 

I know that PHP provides a function `func_get_args()` that returns an array comprising a function's argument list. I've never properly used it before. Could it be applied in this situation?

Reply:
@rgadon107 Let me explain:

```php
$this->assertEquals( 20, has_filter( 'filter_meta_box_custom_fields' , function () {}, 20, 3 ); // returned false.
```

Let's go back to what [`has_filter()`](https://knowthecode.io/docx/wordpress/has_filter). This function requires the following:

- The name of the filter event
- The name of the function/method callback that is being checked, ie Is this callback registered to this filter event?
- Priority level

Notice that the 2nd argument is the exact callback that's registered to the filter event. Hold this in your mind.

A closure (which is also known as anonymous or unnamed function) does not have a user-defined name that is bound to the closure. Right? Why is this important?

Think about how PHP is managing memory. The callables (which includes functions, methods, and closures) are stored in memory. Named (ie the names we give to each callable) callables is stored in the Symbols table in memory and then bound to the memory location where the callable is stored. 

What about unnamed callables, ie closures? There's no name in the Symbols table. That means you can't reference it or call it by a name. Why? It doesn't have one. This distinction is important when you want to do something with that closure in relationship to a filter or action event.

What do I mean? `has_filter()` requires the name or memory location of the callback in order to _look up_ if that callback is registered (hooked) to the filter event. A closure doesn't have a name. It does have a memory location. Huh? Everything in memory has a specific location, which is used by PHP to point to that specific thing in memory. Notice in the Kint printout above that the closure's reference is not a name but instead a long hex code. That's the memory location/ID for that specific unnamed callback (closure).

Now back to your code:

```php
$this->assertEquals( 20, has_filter( 'filter_meta_box_custom_fields' , function () {}, 20, 3 ); // returned false.
```

You have a brand new closure as the 2nd argument. `function() {}` defines a brand new closure. It does _not_ point to the closure that is actually registered. 

## How can you reference a closure?

You have to get the memory ID/location. There are a couple of ways:

1. Storing the closure in a variable.
2. _Discover_ the memory ID/location by looking inside of `$wp_filter` to find the hex code. Messy.

Both of these approaches are messy and unnecessary. If you need to reference the callback, then use a named function instead.

## When to use a closure as a callback

As a general rule, you can use a closure as a callback is when that callback is:

1. Not reusable (not used anywhere else in the code, ie a true callback that is bound to one singular hook)
2. Will not be directly unhooked/unregistered, ie meaning it remains registered/hooked.

Therefore, in general, we say that a closure callback is not usable within `remove_filter()`, `has_filter()`, `remove_action()`, or `has_action()`.